File: Antlr/src/main/scala/io/github/chaosunity/casc/parsing/type/ClassType.scala
package io.github.chaosunity.casc.parsing.`type`

import org.objectweb.asm.Opcodes
import org.objectweb.asm.Opcodes._

class ClassType(val _name: String) extends Type with Opcodes {
    override val name: String = _name

    override val classType: Class[_] = Class.forName(_name)

    override val internalName: String = name.replace(".", "/")

    override val descriptor: String = s"L$internalName;"

    override def loadVariableOpcode: Int = ALOAD

    override def storeVariableOpcode: Int = ASTORE

    override def returnOpcode: Int = ARETURN

    override def addOpcode: Int = throw new RuntimeException(s"Unknown addition operation for class '$name'")

    override def subtractOpcode: Int = throw new RuntimeException(s"Unknown subtraction operation for class '$name'")

    override def multiplyOpcode: Int = throw new RuntimeException(s"Unknown multiplication operation for class '$name'")

    override def divideOpcode: Int = throw new RuntimeException(s"Unknown division operation for class '$name'")
}

File: src/main/kotlin/io/github/chaosunity/casc/parsing/type/ClassType.kt
package io.github.chaosunity.casc.parsing.type

import jdk.internal.org.objectweb.asm.Opcodes.*

class ClassType(override val typeName: String) : Type {
    override fun classType(): Class<*> = try {
        Class.forName(typeName)
    } catch (e: ClassNotFoundException) {
        throw RuntimeException()
    }

    override val internalName: String = typeName.replace('.', '/')
    override val descriptor: String = "L$internalName;"
    override val loadVariableOpcode: Int = ALOAD
    override val storeVariableOpcode: Int = ASTORE
    override val returnOpcode: Int = ARETURN
    override val addOpcode: Int
        get() = throw RuntimeException("Addition operation is not supported for custom objects")
    override val subtractOpcode: Int
        get() = throw RuntimeException("Subtraction operation is not supported for custom objects")
    override val multiplyOpcode: Int
        get() = throw RuntimeException("Multiplication operation is not supported for custom objects")
    override val divideOpcode: Int
        get() = throw RuntimeException("Division operation is not supported for custom objects")

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        val classType = other as ClassType
        return typeName == classType.typeName
    }

    override fun hashCode(): Int {
        return typeName.hashCode()
    }
}